### PR TITLE
Fixed #5609: supported workspace search in dirty (unsaved) file content

### LINKED ISSUE #5609: [search-in-workspace] search does not respect dirty changes
**Description**

The `search-in-workspace` widget does not respect dirty changes.
If an editor `A` has dirty changes and the `search-in-workspace` is used to search such changes, no results are found from `A`.

<div align='center'>

a (2)

</div>

**How to Reproduce**
1. open an editor and make changes (do not save and use a unique string ex: `vincent`).
2. open the `search-in-workspace` widget.
3. search for the unique string, the `search-in-workspace` widget will not show the result.

### PR DESCRIPTION
**Pull request by fangnx**
#### What it does
Fixed #5609.
- The `Search-in-Workspace` now can correctly search content in dirty files (file with unsaved changes).
- Search results in dirty files will be replaced in correct line and character position.
- Implemented by conducting a search in all currently tracked dirty files, before the backend ripgrep search.

#### How to test
- Open editor files in a workspace.
- Make changes in the editors without saving.
- Open `Search-in-Workspace` (`ctrl`+`shift`+`f`) and search for a certain keyword. The results should correctly reflect the current state of the dirty editors. The unsaved content in the editors should be shown in the search results, and properly highlighted in the editors.
- Try with the search options. All the `SearchInWorkspaceOptions` are supported, including:
    - [x] Match Regex (on/off)
    - [x] Match Case (on/off)
    - [x] Match Whole Word (on/off)
    - [x] Include ignored files (on/off)
    - [x] Files to include
    - [x] Files to exclude
    - [x] Maximum number of results
- Do `Replace` or `Replace All`. Matches in the dirty file (files with unsaved changes) should be replaced correctly.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to review in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts [hunk 1]
```diff
@@ -209,7 +209,6 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
                             length,
                             lineText: lineText.replace(/[\r\n]+$/, ''),
                         };
-
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts]: unnecessary change.

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 2]
```diff
@@ -502,7 +612,8 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
     }
 
     protected renderRemoveButton(node: TreeNode): React.ReactNode {
-        return <span className='remove-node' onClick={e => this.remove(node, e)} title='Dismiss'></span>;
+        return <span className='remove-node' onClick={e => this.remove(node, e)
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: This formatting looks a little odd to me and probably shouldn't be part of the changes.

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 3]
```diff
@@ -402,8 +511,9 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
     protected renderReplaceButton(node: TreeNode): React.ReactNode {
         const isResultLineNode = SearchInWorkspaceResultLineNode.is(node);
         return <span className={isResultLineNode ? 'replace-result' : 'replace-all-result'}
-            onClick={e => this.doReplace(node, e)}
-            title={isResultLineNode ? 'Replace' : 'Replace All'}></span>;
+            onClick={e => this.doReplace(node, e)
+            }
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: This formatting looks a little odd to me and probably shouldn't be in the changes.

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 4]
```diff
@@ -235,7 +345,6 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
                 } else {
                     const newRootFolderNode = this.createRootFolderNode(result.root);
                     tree.set(result.root, newRootFolderNode);
-
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: unnecessary change.

File: packages/core/src/browser/shell/application-shell.ts [hunk 5]
```diff
@@ -1405,6 +1405,13 @@ export class ApplicationShell extends Widget {
         return [...this.tracker.widgets];
     }
 
+    /**
+     * Returns all tracked dirty widgets.
+     */
+    get dirtyWidgets(): ReadonlyArray<Widget> {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/shell/application-shell.ts]: In general, the `search-in-workspace` should only care about dirty editors, not necessarily all dirty widgets.
Later when calling the method, no filter is performed to extract `Editors`.

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 6]
```diff
@@ -196,6 +197,106 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
         });
     }
 
+    getRoot(filePath: string, rootUris: string[]): URI {
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: Can you provide documentation for this method?

**Comment by vince-fugnitto** [comment]
If I have a search result from my dirty editor with multiple matches, only a single match is displayed in the tree for that line. Here I'd expect 3 matches:

<div align='center'>

Screen Shot 2019-07-05 at 4 06 34 PM


</div>

**Comment by fangnx** [comment]
> If I have a search result from my dirty editor with multiple matches, only a single match is displayed in the tree for that line. Here I'd expect 3 matches:
> 
> Screen Shot 2019-07-05 at 4 06 34 PM

I will have this fixed.

File: packages/core/src/browser/shell/application-shell.ts (hunk 5 above)
Comment (vince-fugnitto) [reply packages/core/src/browser/shell/application-shell.ts]: I think maybe using the `editor-manager.all` method and extracting dirty editors may be sufficient for the implementation of the PR.

Comment (fangnx) [reply packages/core/src/browser/shell/application-shell.ts]: Just fixed.

**Comment by lmcbout** [comment]
@fangnx 
When the search is perform and the user keep modifying a file before triggering the "replace all",  you are corrupting the dirty file.
@ options:
1- Just before you trigger the "replace all" command, you need to refresh the "search" result
2- Or when you edit the file, you need an event to update the search result for the selected file before you can trigger the "replace all" command, to make sure the search result is up to date

I prefer option 2 because you maintain a search result up-to-date . 
Note: To maintain the search result up to date, you need to search only the dirty file and refresh only for that file, no need to perform the search for the whole workspace all the times.

**Comment by vince-fugnitto** [comment]
@lmcbout everything you mentioned, is it not the same issue as https://github.com/theia-ide/theia/issues/2319?
That behavior is part of master and not this PR (it also happens for non-dirty files).

**Comment by lmcbout** [comment]
@vince-fugnitto (option 2 suggested)
Yes, implementing the events  for issue #2319 could solved it, but this is not an enhancement, it is a bug.
It becomes very visible when testing this PR :)

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 7]
```diff
@@ -39,10 +39,11 @@ import { WorkspaceService } from '@theia/workspace/lib/browser';
 import { FileResourceResolver } from '@theia/filesystem/lib/browser';
 import { SearchInWorkspaceResult, SearchInWorkspaceOptions } from '../common/search-in-workspace-interface';
 import { SearchInWorkspaceService } from './search-in-workspace-service';
+import { SearchInWorkspacePreferences } from './search-in-workspace-preferences';
 import { MEMORY_TEXT } from './in-memory-text-resource';
+import { FileUri } from '@theia/core/lib/node/file-uri';
```
Comment (marechal-p) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: You are importing a library intended for `node` in the `browser`: that's a big no-no.

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 8]
```diff
@@ -196,9 +198,134 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
...
+    /**
+     * Returns the root folder URI that a file belongs to.
+     * In the case that a file belongs to more than one root folders, returns the root folder closest to the file.
+     * If the file is not from the current workspace, returns an empty URI.
+     * @param {string} filePath - path of the file.
+     * @param {stirng[]} rootUris - URIs of the root folders in the workspace
+     * @returns URI of the root folder.
+     */
+    getRoot(filePath: string, rootUris: string[]): URI {
+        const roots = rootUris.filter(root => new URI(root).withScheme('file').isEqualOrParent(FileUri.create(filePath).withScheme('file')));
```
Comment (marechal-p) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: The `FileUri` usage is problematic here, you may have to try using a different API, see https://github.com/theia-ide/theia/wiki/Coding-Guidelines#uripath

Comment (fangnx) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: I will change the part related to URI. Thanks for pointing it out!

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 9]
```diff
@@ -196,9 +198,134 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
...
+        }
+        return new URI();
+    }
+
+    /**
+     * Returns all matches in a dirty file.
+     */
+    findMatches(searchTerm: string, fileContent: string): { lineNumber: number, character: number, lineText: string }[] {
+        const matches = [];
+        const allLines = fileContent.split('\n');
```
[the remark this answers is not in the document]
Comment (fangnx) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: Yeah it should be this ;)

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 10]
```diff
@@ -207,39 +334,20 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
             this.refreshModelChildren();
             return;
         }
+
+        // Searches in the dirty editors first.
+        this.searchInDirtyFiles(searchTerm);
```
Comment (svenefftinge) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: we should respect all search options for dirty file search as well

Comment (fangnx) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: Yes I just realized it. I am working on the support for search options right now ;)

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 11]
```diff
@@ -196,9 +198,134 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
...
+    }
+
+    /**
+     * Seaches in all files in the current workspace.
+     */
     async search(searchTerm: string, searchOptions: SearchInWorkspaceOptions): Promise<void> {
         this.searchTerm = searchTerm;
-        const collapseValue: string = this.searchInWorkspacePreferences['search.collapseResults'];
+        // Stores URIs of the dirty editors to avoid duplicated search results.
+        this.dirtyFileUris = new Set<string>();
```
[the remark this answers is not in the document]
Comment (fangnx) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: Fixed.

**Comment by fangnx** [comment]
The `SearchInWorkspaceOptions' are all supported in dirty search except for `includeIgnored`, which is hard to implement as `search-in-workspace` does not have `scm` as its dependency. I think it is a minor issue it is rarely the case that a user would make an ignored file dirty and would like to exclude that file in global search.

**Comment by vince-fugnitto** [comment]
> The `SearchInWorkspaceOptions` are all supported in dirty search except for `includeIgnored`, which is hard to implement as `search-in-workspace` does not have `scm` as its dependency. I think it is a minor issue as it is rarely the case that a user would make an ignored file dirty and would like to exclude that file in global search.

I'm not sure it needs the `scm` extension, it's functionality provided by the `filesystem`.
The filesystem also has the preference `files.exclude` which defines glob patterns for excluding files and folders.

**Comment by marechal-p** [comment]
@vince-fugnitto in this case `includeIgnored` refers to "gitignored"-or-so files.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto in this case `includeIgnored` refers to "gitignored"-or-so files.

In VSCode I am able to set my `files.exclude` preference and ignore the globs I set despite no `.gitignore` file being present in my workspace.

<div align='center'>


image


</div>



<div align='center'>

image


</div>

**Comment by vince-fugnitto** [comment]
@marechal-p I assume its a combination of `.gitignore` and settings.

**Comment by marechal-p** [comment]
It might indeed, didn't know VS Code combined the two.

**Comment by fangnx** [comment]
@vince-fugnitto @marechal-p All the search options have been supported ;)

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 12]
```diff
@@ -196,9 +198,134 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
...
+     * @param {string} filePath - path of the file.
+     * @param {stirng[]} rootUris - URIs of the root folders in the workspace
+     * @returns URI of the root folder.
+     */
+    getRoot(filePath: string, rootUris: string[]): URI {
+        const roots = rootUris.filter(root => new URI(root).withScheme('file').isEqualOrParent(FileUri.create(filePath).withScheme('file')));
+        if (roots.length > 0) {
+            return FileUri.create(FileUri.fsPath(roots.sort((r1, r2) => r2.length - r1.length)[0]));
+        }
+        return new URI();
```
[the remark this answers is not in the document]
Comment (vince-fugnitto) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: I believe it's done on purpose to follow that of `ripgrep-search-in-workpsace-server.ts`

https://github.com/theia-ide/theia/blob/bd044245d03583949ec2026e274819094b0ca659/packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts#L250-L256

**Comment by vince-fugnitto** [comment]
@fangnx only other comment I have regarding the code is if it'd be possible to add test cases for the custom searching we do to mimic ripgrep. I'd be more comfortable if there was adequate tests to handle the new dirty searching.

**Comment by lmcbout** [comment]
@vince-fugnitto Any reasons why we this PR is not merged !

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Any reasons why we this PR is not merged !

There's a conflict?

**Comment by vince-fugnitto** [comment]
@lmcbout I updated the PR (resolving the conflicts)

**Comment by lmcbout** [comment]
@akosyakov Do you have any objections if we merge this PR ?

File: packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx [hunk 13]
```diff
@@ -198,9 +201,175 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
...
+            dirtyWidgets = dirtyWidgets.filter(widget => !excludedPatterns.some(pattern => minimatch(widget.title.label, pattern)));
+        }
+
+        let numberOfResults = 0;
+        dirtyWidgets.forEach(async w => {
+            const fileUri: string = w.editor.uri.toString();
+            const roots = await this.workspaceService.roots;
+            const root: string = this.getRoot(w.editor.uri.path.toString(), roots.map(r => r.uri)).toString();
+            const fileContent: string = w.editor.document.getText();
+            const matches = this.findMatches(searchTerm, fileContent, searchOptions);
```
Comment (akosyakov) [inline packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: monaco model has API to search matches: https://github.com/microsoft/monaco-editor/blob/0a4ed9a829ce6deaf794ea3dfaf4e814a4c10120/monaco.d.ts#L1646-L1669

You don't need to implement it.

Comment (vince-fugnitto) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: That sounds much better, I assume it requires waiting on #5901

Comment (akosyakov) [reply packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx]: no, it should be there in Monaco 0.14 as well
